### PR TITLE
password reset using mnemonic

### PR DESCRIPTION
**Pull request by interestIngc**
## Overview
This PR contains a Typescript module ```mnemonic``` (under server/mnemonic) with implementation of methods to generate and validate ```bip39``` mnemonic phrases, as well as to derive a private / public key pair from it with a use of ```bip32``` protocol.

```user_portal/register.html``` now contains logic to generate a mnemonic when user gets registered, to derive a public key from this mnemonic and to transmit this public key to the backend.

```reset-password-page.html``` contains frontend logic to execute the password reset procedure, including validating the mnemonic, deriving a private key, signing a message and transmitting the obtained signature to the backend.

Two new API endpoints are added:
1. ```/reset-password-page``` -- returns ```reset-password-page.html```
2. ```/api/v1/reset-password``` -- looks up user's public key from the db based on the username, validates the signature using this public key, and updates user's password in the db if validation is successful.

### SECTION 1: RESOURCE SERVER (http://localhost:5001/)
- [x] Created a new user on the Layer8 Portal

### SECTION 2: WE'VE GOT POEMS (http://localhost:5173/)
- [x] Login as 'tester', '1234'
- [x] Choose login with layer8 
- [x] Logging in with the newly registered credentials from Section 1 succeeds in popup
- [x] User chooses to share their new "Username" & "Country" from the Layer8 Resource Server
- [x] Choose poems 1 - 3 works correctly 

### SECTION 3: IMSHARER (http://localhost:5174/)
- [x] Uploads an image works
- [x] Reload leads to instant reload (demonstrating proper caching)

### CODE REVIEW

File: server/resource_server/repository/mem_repository.go [hunk 1]
```diff
@@ -408,3 +408,11 @@ func (r *MemoryRepository) SaveZkSnarksKeyPair(keyPair models.ZkSnarksKeyPair) (
 func (r *MemoryRepository) GetLatestZkSnarksKeys() (models.ZkSnarksKeyPair, error) {
 	return models.ZkSnarksKeyPair{}, nil
 }
+
+func (r *MemoryRepository) GetUserForUsername(username string) (models.User, error) {
+	return models.User{}, nil
```
Comment (stravid87) [inline server/resource_server/repository/mem_repository.go:413]: Nice work around. I was wondering how you'd get around this.

**Review by stravid87 — COMMENTED** [review]
Assuming I succeed in running this locally, then this PR is approved. 
Well done. Very thorough. I thought I found a place where you didn't handle the errors but in the end, the code was in fact there. We should check out why that one test case is not passing? And I'll run it locally as well. But, from what I see, this looks great.

**Comment by stravid87** [comment]
Ran perfectly on the first run. Fixing the tests here is all that is necessary in my opinion now before merge: `..\layer8\test\cypress\cypress\e2e\client_and_user_portal\user_portal.cy.js`

File: server/assets-v1/templates/src/pages/user_portal/password_reset/reset-password-page.html [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+                    "[[ .ProxyURL ]]/api/v1/reset-password",
+                    {
+                        method: "POST",
+                        headers: {
+                            "Content-Type": "application/json",
+                        },
+                        body: JSON.stringify({
+                            username: username.value,
+                            signature: Array.from(signature),
+                            new_password: newPassword.value
```
[the remark this answers is not in the document]
Comment (interestIngc) [reply server/assets-v1/templates/src/pages/user_portal/password_reset/reset-password-page.html:143]: Yeah, we'll be working on that!

File: server/assets-v1/templates/src/pages/user_portal/register.html [hunk 3]
```diff
@@ -4,131 +4,177 @@
     <meta charset="UTF-8" />
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
     <link
-      rel="stylesheet"
-      href="/assets-v1/templates/assets/styles/output.css"
+        rel="stylesheet"
+        href="/assets-v1/templates/assets/styles/modal.css"
```
[the remark this answers is not in the document]
Comment (interestIngc) [reply server/assets-v1/templates/src/pages/user_portal/register.html:8]: There might be. Shall we investigate this or just merge it as it is?

File: server/resource_server/repository/mem_repository.go (hunk 1 above)
Comment (interestIngc) [reply server/resource_server/repository/mem_repository.go:413]: Yeah, since this memory repository is legacy, we can just stub the new repository methods

**Pull request merged**